### PR TITLE
fix saving of black display names etc. of items in config

### PR DESCRIPTION
**Pull request by weaondara**
this fixes saving of items with black display names in yaml configs

### CODE REVIEW

File: Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
++            // Paper end
+             if (modi.isBold()) {
+                 out.append(EnumChatFormat.BOLD);
+             }
+@@ -192,9 +202,9 @@ public final class CraftChatMessage {
+             }
+             out.append(c.getText());
+         }
+-        return out.toString().replaceFirst("^(" + defaultColor + ")*", "");
++        return out.toString(); // Paper - fix deletion of color codes at the beginning of result string (old behavior deleted black color codes in display names of items)
```
Comment (kickash32) [inline Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: missing Paper comment

File: Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     }
+-
++
```
Comment (kickash32) [inline Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: Unnecessary diff

**Review by kickash32 — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch (hunk 1 above)
Comment (weaondara) [inline Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: where did i miss it?

Comment (kickash32) [reply Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: never mind I can't read...

File: Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch (hunk 2 above)
Comment (weaondara) [reply Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: ok fixed that but it says branches are diverging
should i do a force push or is there another option
i used the 'paper edit server'

Comment (kickash32) [reply Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: force pushing to your branch should be fine

Comment (weaondara) [reply Spigot-Server-Patches/0422-Fix-saving-of-black-display-names-etc.-of-items-in-c.patch]: ok thx

**Comment by Spottedleaf** [comment]
It looks to me like if it's deleting the default colour at the start and that's an issue, then the default colour code is actually wrong?

**Comment by weaondara** [comment]
the problem with old behavior is that no matter what default color you use there is a case where it changes the displayname. some examples for default colors:
black: displayname '§0test' becomes 'test' -> broken
white: displayname '§ftest' becomes 'test' and the client show the item name in italic -> broken
reset: displayname '§rtest' becomes 'test' -> broken
the patch i proposed ensures that the color code at the beginning is not erased if it matches the default color

**Comment by Mystiflow** [comment]
For '&aTest &0Item' as an example wouldn't the 'Item' be broken still as it's the default colour?

**Comment by weaondara** [comment]
any news on this?

**Comment by electronicboy** [comment]
My only real major concern with such a patch is compatibility, we've had issues in the past with plugin compatibility when trying to address an issue similar to this

**Comment by weaondara** [comment]
What exactly do you mean with compatibility issues? Currently the issue is that if you serialize an item with a black name, that Color Code gets removed because of spigots conversion: https://hub.spigotmc.org/jira/browse/SPIGOT-4441
For us that is causing issues, when a player has an item with a black name in the inventory, and you serialize that inventory, the item looses its black name.
If you want to talk about that in more detail you can write me on discord, Alice#9276

**Comment by aurorasmiles** [comment]
I think this can be closed as well with https://github.com/PaperMC/Paper/commit/c3c005c947e8595dcf9a8fa765ffad41c22099c7

**Comment by aikar** [comment]
Reopening and merging because Spigot didn't fix this issue, they made it 1000x worse, and this PR fixes it for all the now broken cases.

Verified the behavior in this PR is the correct logic that should of existed.
